### ISSUE TITLE
[FEA] C++ cuGraph library handle

### ISSUE THREAD

**Issue by afender** [rapidsai/cugraph#815]
There are many reasons why we may want a handle :
- communicator
- Keep track of library wide events (status)
- Counters (devices, streams, calls, … )
- Stream(s)
- Allocator
- Cuda library handles (expensive to create)

We managed to postpone this so far.
It becomes pretty much required in the OPG paradigm though. We need to pass the communicator so that we have a constant view of the multi-GPU / multi-process context between calls

**Comment by afender** [comment]
the handle is now in RAFT : https://github.com/rapidsai/raft/pull/3

**Comment by afender** [comment]
The graph class has been updated and multi-GPU functions accept the handle.
We now need a consistent solution so that the handle content can be accessed from every feature.

Some of the challenges are:
- avoid undefined behavior when the graph has a different handle than the algo
- prevent recreating handle(s) as this can be costly

This task unblock stream support #957
